menu.py

```diff
@@ -1,38 +1,452 @@
+    kVK_UpArrow: "up",
+    # kVK_ISO_Section: '',
+    # kVK_JIS_Yen: '',
+    # kVK_JIS_Underscore: '',
+    # kVK_JIS_KeypadComma: '',
+    # kVK_JIS_Eisu: '',
+    # kVK_JIS_Kana: '',
+}
+
+
+def active_menu_bar():
     return ui.active_app().children.find_one(AXRole="AXMenuBar", max_depth=0)
 
 
+def selected_menu_and_path():
+    """Returns selected element in menu bar, and path to it"""
+    selected_menu = active_menu_bar()
+    menu_path = []
+    while True:
+        if not selected_menu.AXSelectedChildren:
+            break
+        selected_menu = selected_menu.AXSelectedChildren[0]
+
+        if not (title := selected_menu.get("AXTitle")):
+            return selected_menu, menu_path
+
+        menu_path.append(selected_menu.AXTitle)
+        # XXX(nriley) this is True even if list empty - report Talon bug? children also doesn't support len()
+        # print(selected_menu.children, bool(selected_menu.children))
+        try:
+            selected_menu = selected_menu.children.find_one(
+                AXRole="AXMenu", max_depth=0
+            )
+        except ui.UIErr:
+            break
+
+    return selected_menu, menu_path
+
+
+def mouse_pos_menu_and_path():
+    """Returns element in menu bar under cursor, and path to it"""
+    menu_path = []
+    element_at_mouse_pos = ui.element_at(*ctrl.mouse_pos())
+    if element_at_mouse_pos.AXRole in ("AXMenu", "AXMenuItem", "AXMenuBarItem"):
+        menu = element_at_mouse_pos
+        while True:
+            if menu.AXRole in ("AXMenuItem", "AXMenuBarItem"):
+                menu_path.append(menu.AXTitle)
+            menu = menu.AXParent
+            if menu.AXRole == "AXMenu":
+                menu = menu.AXParent
+            elif menu.AXRole == "AXMenuBar":
+                menu_path.reverse()
+                break
+        else:  # not in menu bar
+            return None, []
+
+    return element_at_mouse_pos, menu_path
+
+
+def selected_menu_path_strategy():
+    """Returns 'selected' element in menu bar, path to it and strategy used to find it"""
+    selected_menu, selected_menu_path = selected_menu_and_path()
+
+    if (
+        selected_menu_path
+        and hasattr(ui, "element_at")
+        and selected_menu.AXRole != "AXMenuItem"
+    ):
+        element_at_mouse_pos, mouse_pos_menu_path = mouse_pos_menu_and_path()
+        if element_at_mouse_pos and len(mouse_pos_menu_path) > len(selected_menu_path):
+            return (
+                element_at_mouse_pos,
+                mouse_pos_menu_path,
+                "Found under mouse pointer",
+            )
+
+    return selected_menu, selected_menu_path, "Found selected"
+
+
+def selected_menu_key_path_strategy():
+    """Returns Talon-format key equivalent of the menu item that is currently highlighted,
+    , path to it and strategy used to find it"""
+    selected_menu, menu_path, strategy = selected_menu_path_strategy()
+
+    if (not selected_menu) or selected_menu.AXRole != "AXMenuItem":
```

Comment:
Are these parentheses necessary?

Reply:
Just there for readability.

Reply:
imo, if they aren't needed to force operator precedence, I think they detract from readability (the reader has to to wonder why they're there)

Reply:
I spelled this a bit differently (`selected_menu is None`) which is easier to read without parens...

Reply:
sure, lgtm